### PR TITLE
[DEV-920] Reward payout process

### PR DESCRIPTION
**Pull request by TonySemikin**
- [x] wait until feature/DEV-922-reward-strategies is merged to this PR
- [x] add environment variables -> rew_wallet_address

### CODE REVIEW

File: src/blockchain/ain/utils/defichain.util.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+  async getAvailableTokenAmount(assetName: string, client: DeFiClient): Promise<number> {
+    const tokens = await client.getToken();
+    const token = tokens
+      .filter((t) => t.owner === Config.blockchain.default.rewWalletAddress)
+      .map((t) => client.parseAmount(t.amount))
+      .find((pt) => pt.asset === assetName);
+
+    return token ? token.amount : 0;
+  }
```
[the remark this answers is not in the document]
Comment (TonySemikin) [reply src/blockchain/ain/utils/defichain.util.ts:34]: use wallet address as a parameter

File: src/subdomains/analytics/application/mappers/compact-history-dto.mapper.ts [hunk 2]
```diff
@@ -14,6 +14,11 @@ export class CompactHistoryDtoMapper {
     [WithdrawalStatus.PAYING_OUT]: CompactHistoryStatus.PENDING,
     [WithdrawalStatus.CONFIRMED]: CompactHistoryStatus.CONFIRMED,
     [WithdrawalStatus.FAILED]: CompactHistoryStatus.FAILED,
+    [RewardStatus.CREATED]: null,
+    [RewardStatus.PAUSED]: CompactHistoryStatus.PENDING,
+    [RewardStatus.PREPARATION_PENDING]: CompactHistoryStatus.PENDING,
+    [RewardStatus.PREPARATION_CONFIRMED]: CompactHistoryStatus.PENDING,
+    [RewardStatus.CONFIRMED]: CompactHistoryStatus.CONFIRMED,
```
[the remark this answers is not in the document]
Comment (TonySemikin) [reply src/subdomains/analytics/application/mappers/compact-history-dto.mapper.ts:21]: remove [RewardStatus.CONFIRMED]: CompactHistoryStatus.CONFIRMED

File: src/subdomains/staking/domain/entities/staking.entity.ts [hunk 3]
```diff
@@ -54,6 +54,7 @@ export class Staking extends IEntity {
   @OneToMany(() => Withdrawal, (withdrawal) => withdrawal.staking, { eager: true, cascade: true })
   withdrawals: Withdrawal[];
 
+  // TODO -> this is not needed, just send to whatever provided in reward, this is a BlockchainAddress -> RewardBlockchainAddress
   @ManyToOne(() => WalletBlockchainAddress, { eager: true, nullable: false })
   rewardsPayoutAddress: WalletBlockchainAddress;
```
[the remark this answers is not in the document]
Comment (TonySemikin) [reply src/subdomains/staking/domain/entities/staking.entity.ts:59]: I'll keep it in this PR, already removed in next one

File: src/subdomains/staking/domain/entities/reward.entity.ts [hunk 4]
```diff
@@ -1,74 +1,143 @@
...
+
+  @Column({ type: 'float', nullable: false, default: 0 })
+  feeAmount: number;
+
+  @ManyToOne(() => Asset, { eager: true, nullable: false })
+  targetAsset: Asset;
 
   @Column({ nullable: true })
-  reinvestTxId: string;
+  targetAddress: string;
```
[the remark this answers is not in the document]
Comment (TonySemikin) [reply src/subdomains/staking/domain/entities/reward.entity.ts:43]: this is already deleted in followup PR

File: src/subdomains/staking/application/services/staking-reward-dex.service.ts [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    return this.rewardRepo
+      .createQueryBuilder('reward')
+      .leftJoin('reward.referenceAsset', 'referenceAsset')
+      .select('SUM(outputReferenceAmount)', 'amount')
+      .where('status = :status', { status: RewardStatus.CREATED })
+      .andWhere('referenceAsset.name = :name', { name: 'DFI' })
+      .andWhere('referenceAsset.blockchain = :blockchain', { blockchain: Blockchain.DEFICHAIN })
+      .getRawOne<{ amount: number }>()
+      .then((r) => r.amount);
+  }
```
[the remark this answers is not in the document]
Comment (TonySemikin) [reply src/subdomains/staking/application/services/staking-reward-dex.service.ts:87]: I'll do it for all the codebase in my refactoring PR

File: migration/1669638666343-rewardsPayout.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        await queryRunner.query(`ALTER TABLE "staking_analytics" DROP COLUMN "operatorCount"`);
+        await queryRunner.query(`ALTER TABLE "staking_analytics" DROP COLUMN "tvl"`);
+        await queryRunner.query(`ALTER TABLE "reward" DROP CONSTRAINT "DF_903071fa6f845566a03224c6555"`);
+        await queryRunner.query(`ALTER TABLE "reward" DROP COLUMN "amount"`);
+        await queryRunner.query(`ALTER TABLE "reward" DROP COLUMN "reinvestTxId"`);
+        await queryRunner.query(`ALTER TABLE "reward" DROP COLUMN "reinvestOutputDate"`);
+        await queryRunner.query(`ALTER TABLE "reward" DROP CONSTRAINT "DF_bee3de89dbf732f919cbc184b5e"`);
+        await queryRunner.query(`ALTER TABLE "reward" DROP COLUMN "fee"`);
+        await queryRunner.query(`ALTER TABLE "reward" DROP COLUMN "assetId"`);
```
[the remark this answers is not in the document]
Comment (TonySemikin) [reply migration/1669638666343-rewardsPayout.js:19]: removed migrations until feature/DEV-922-reward-strategies is merged to this PR

**Review by davidleomay — APPROVED** [review]

**Pull request merged**